File: apps/ios/SocialApp/APIClient.swift
import Foundation
import Security

struct TokenData: Codable {
    let accessToken: String
    let refreshToken: String
    let expiresIn: Int

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
        case expiresIn = "expires_in"
    }
}

struct LoginResponse: Codable {
    let code: Int
    let message: String
    let langKey: String
    let data: TokenData?

    enum CodingKeys: String, CodingKey {
        case code, message, data
        case langKey = "lang_key"
    }
}

struct PostItem: Codable, Identifiable {
    let id: Int64
    let content: String
    let likeCount: Int
    let commentCount: Int

    enum CodingKeys: String, CodingKey {
        case id, content
        case likeCount = "like_count"
        case commentCount = "comment_count"
    }
}

struct RelationData: Codable {
    let isFollowing: Bool
    let followerCount: Int
    let followingCount: Int

    enum CodingKeys: String, CodingKey {
        case isFollowing = "is_following"
        case followerCount = "follower_count"
        case followingCount = "following_count"
    }
}

struct UserBrief: Codable {
    let id: Int64
    let nickname: String
}

struct NotificationItem: Codable, Identifiable {
    let id: Int64
    let type: String
    let content: String
    let read: Bool
    let actor: UserBrief?

    enum CodingKeys: String, CodingKey {
        case id, type, content, read, actor
    }
}

enum Keychain {
    private static let service = "com.social.app"

    static func save(_ value: String, forKey key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecValueData as String: Data(value.utf8),
        ]
        SecItemDelete(query as CFDictionary)
        SecItemAdd(query as CFDictionary, nil)
    }

    static func load(_ key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
              let data = item as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }
}

final class APIClient {
    static let shared = APIClient()
    private let base = "http://127.0.0.1:8787" // 模拟器访问宿主机
    private let session = URLSession.shared
    var accessToken: String = ""

    private init() {
        accessToken = Keychain.load("access_token") ?? ""
    }

    func request(_ method: String, _ path: String, body: [String: Any]? = nil,
                 completion: @escaping (Data?, Error?) -> Void) {
        var req = URLRequest(url: URL(string: base + path)!)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if !accessToken.isEmpty {
            req.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        }
        if let body = body {
            req.httpBody = try? JSONSerialization.data(withJSONObject: body)
        }
        session.dataTask(with: req) { data, _, error in
            completion(data, error)
        }.resume()
    }

    func login(email: String, password: String, completion: @escaping (LoginResponse?) -> Void) {
        request("POST", "/api/v1/auth/login", body: ["email": email, "password": password]) { data, _ in
            guard let data = data else { completion(nil); return }
            let resp = try? JSONDecoder().decode(LoginResponse.self, from: data)
            resp?.data.map {
                self.accessToken = $0.accessToken
                Keychain.save($0.accessToken, forKey: "access_token")
            }
            completion(resp)
        }
    }

    func register(email: String, password: String, nickname: String, completion: @escaping (LoginResponse?) -> Void) {
        request("POST", "/api/v1/auth/register",
                body: ["email": email, "password": password, "nickname": nickname]) { data, _ in
            guard let data = data else { completion(nil); return }
            let resp = try? JSONDecoder().decode(LoginResponse.self, from: data)
            resp?.data.map {
                self.accessToken = $0.accessToken
                Keychain.save($0.accessToken, forKey: "access_token")
            }
            completion(resp)
        }
    }

    func timeline(completion: @escaping ([PostItem]) -> Void) {
        request("GET", "/api/v1/posts") { data, _ in
            guard let data = data,
                  let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let d = root["data"] as? [String: Any],
                  let list = d["list"] else { completion([]); return }
            let listData = try? JSONSerialization.data(withJSONObject: list)
            let items = listData.flatMap { try? JSONDecoder().decode([PostItem].self, from: $0) } ?? []
            completion(items)
        }
    }

    func createPost(content: String, completion: @escaping (Bool) -> Void) {
        request("POST", "/api/v1/posts", body: ["content": content]) { data, _ in
            guard let data = data,
                  let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                completion(false); return
            }
            completion((root["code"] as? Int) == 0)
        }
    }

    func me(completion: @escaping (Int64?) -> Void) {
        request("GET", "/api/v1/me") { data, _ in
            completion(self.dataDict(data)?["user_id"] as? Int64)
        }
    }

    func follow(id: Int64, completion: @escaping (Bool) -> Void) {
        request("POST", "/api/v1/users/\(id)/follow") { data, _ in completion(self.ok(data)) }
    }

    func unfollow(id: Int64, completion: @escaping (Bool) -> Void) {
        request("POST", "/api/v1/users/\(id)/unfollow") { data, _ in completion(self.ok(data)) }
    }

    func relation(id: Int64, completion: @escaping (RelationData?) -> Void) {
        request("GET", "/api/v1/users/\(id)/relation") { data, _ in
            guard let d = self.dataDict(data),
                  let json = try? JSONSerialization.data(withJSONObject: d) else { completion(nil); return }
            completion(try? JSONDecoder().decode(RelationData.self, from: json))
        }
    }

    func notifications(completion: @escaping ([NotificationItem]) -> Void) {
        request("GET", "/api/v1/notifications") { data, _ in
            guard let d = self.dataDict(data),
                  let list = d["list"],
                  let json = try? JSONSerialization.data(withJSONObject: list) else { completion([]); return }
            completion((try? JSONDecoder().decode([NotificationItem].self, from: json)) ?? [])
        }
    }

    func unreadCount(completion: @escaping (Int) -> Void) {
        request("GET", "/api/v1/notifications/unread-count") { data, _ in
            completion(self.dataDict(data)?["unread_count"] as? Int ?? 0)
        }
    }

    func markAllRead(completion: @escaping (Bool) -> Void) {
        request("POST", "/api/v1/notifications/read-all") { data, _ in completion(self.ok(data)) }
    }

    private func dataDict(_ data: Data?) -> [String: Any]? {
        guard let data = data,
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        return root["data"] as? [String: Any]
    }

    private func ok(_ data: Data?) -> Bool {
        guard let data = data,
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return false }
        return (root["code"] as? Int) == 0
    }
}

File: apps/android/app/src/main/java/com/social/app/api/ApiClient.kt
package com.social.app.api

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

private const val BASE = "http://10.0.2.2:8788" // 模拟器访问宿主机

@Serializable
data class LoginResponse(
    val code: Int,
    val message: String = "",
    @kotlinx.serialization.SerialName("lang_key") val langKey: String = "",
    val data: TokenData? = null,
)

@Serializable
data class TokenData(
    @kotlinx.serialization.SerialName("access_token") val accessToken: String,
    @kotlinx.serialization.SerialName("refresh_token") val refreshToken: String,
    @kotlinx.serialization.SerialName("expires_in") val expiresIn: Long,
)

@Serializable
data class PostItem(
    val id: Long,
    val content: String,
    @kotlinx.serialization.SerialName("like_count") val likeCount: Int = 0,
    @kotlinx.serialization.SerialName("comment_count") val commentCount: Int = 0,
    @kotlinx.serialization.SerialName("created_at") val createdAt: String = "",
)

@Serializable
data class UserBriefItem(
    val id: Long,
    val nickname: String = "",
    val avatar: String = "",
    val bio: String = "",
    val gender: Int = 0,
)

@Serializable
data class RelationData(
    @kotlinx.serialization.SerialName("is_following") val isFollowing: Boolean,
    @kotlinx.serialization.SerialName("follower_count") val followerCount: Long,
    @kotlinx.serialization.SerialName("following_count") val followingCount: Long,
)

@Serializable
data class NotificationItem(
    val id: Long,
    val type: String,
    @kotlinx.serialization.SerialName("ref_type") val refType: String = "",
    @kotlinx.serialization.SerialName("ref_id") val refId: Long = 0,
    val content: String = "",
    val read: Boolean = false,
    @kotlinx.serialization.SerialName("created_at") val createdAt: String = "",
    val actor: UserBriefItem? = null,
)

@Serializable
data class MeData(val user: MeUser? = null, val profile: ProfileItem? = null)

@Serializable
data class MeUser(val id: Long)

@Serializable
data class ProfileItem(
    val nickname: String = "",
    val avatar: String = "",
    val bio: String = "",
    val gender: Int = 0,
)

object ApiClient {
    private val json = Json { ignoreUnknownKeys = true }
    private val client = OkHttpClient.Builder().build()

    var accessToken: String = ""

    private fun baseRequest(method: String, path: String, body: String? = null): Request.Builder {
        val rb = Request.Builder()
            .url(BASE + path)
            .method(method, body?.toRequestBody("application/json".toMediaType()))
        if (accessToken.isNotEmpty()) rb.header("Authorization", "Bearer $accessToken")
        return rb
    }

    fun login(email: String, password: String): LoginResponse {
        val jsonBody = kotlinx.serialization.json.buildJsonObject {
            put("email", email); put("password", password)
        }.toString()
        val resp = client.newCall(baseRequest("POST", "/api/v1/auth/login", jsonBody).build()).execute()
        val parsed = json.decodeFromString<LoginResponse>(resp.body!!.string())
        parsed.data?.let { accessToken = it.accessToken }
        return parsed
    }

    fun register(email: String, password: String, nickname: String): LoginResponse {
        val jsonBody = kotlinx.serialization.json.buildJsonObject {
            put("email", email); put("password", password); put("nickname", nickname)
        }.toString()
        val resp = client.newCall(baseRequest("POST", "/api/v1/auth/register", jsonBody).build()).execute()
        val parsed = json.decodeFromString<LoginResponse>(resp.body!!.string())
        parsed.data?.let { accessToken = it.accessToken }
        return parsed
    }

    fun timeline(): List<PostItem> =
        getList("/api/v1/posts").map { json.decodeFromJsonElement(PostItem.serializer(), it) }

    fun createPost(content: String): Boolean {
        val jsonBody = kotlinx.serialization.json.buildJsonObject { put("content", content) }.toString()
        return post("/api/v1/posts", jsonBody)
    }

    fun follow(userId: Long): Boolean = post("/api/v1/users/$userId/follow")

    fun unfollow(userId: Long): Boolean = post("/api/v1/users/$userId/unfollow")

    fun relation(userId: Long): RelationData {
        val root = json.parseToJsonElement(get("/api/v1/users/$userId/relation").body!!.string()).jsonObject
        return json.decodeFromJsonElement(RelationData.serializer(), root["data"]!!)
    }

    fun notifications(page: Int = 1): List<NotificationItem> =
        getList("/api/v1/notifications?page=$page").map { json.decodeFromJsonElement(NotificationItem.serializer(), it) }

    fun unreadCount(): Int {
        val root = json.parseToJsonElement(get("/api/v1/notifications/unread-count").body!!.string()).jsonObject
        return root["data"]!!.jsonObject["unread_count"]!!.jsonPrimitive.int
    }

    fun markAllRead(): Boolean = post("/api/v1/notifications/read-all")

    fun me(): Pair<Long, ProfileItem> {
        val data = json.parseToJsonElement(get("/api/v1/auth/me").body!!.string()).jsonObject["data"]!!.jsonObject
        val user = json.decodeFromJsonElement(MeUser.serializer(), data["user"]!!)
        val profile = json.decodeFromJsonElement(ProfileItem.serializer(), data["profile"]!!)
        return user.id to profile
    }

    private fun get(path: String): okhttp3.Response =
        client.newCall(baseRequest("GET", path).build()).execute()

    private fun getList(path: String): List<JsonElement> {
        val root = json.parseToJsonElement(get(path).body!!.string()).jsonObject
        return root["data"]!!.jsonObject["list"]!!.jsonArray.toList()
    }

    private fun post(path: String, body: String? = null): Boolean {
        val resp = client.newCall(baseRequest("POST", path, body).build()).execute()
        return resp.isSuccessful && json.parseToJsonElement(resp.body!!.string()).jsonObject["code"]!!.jsonPrimitive.int == 0
    }
}
